### PR TITLE
[dagster] pipeline presets

### LINKED ISSUE #1145: Adding environments to repository.yml and then supporting in dagit
Per discussion at lunch, add the ability to register environments in repository.yml and then surface them nicely in dagit.

### PR DESCRIPTION
**Pull request by alangenfeld**
Adds the notion of pipeline presets. These are defined in repository.yaml and allow you to specify different "presets" containing a solid subset and environment files. 

resolves #1145

### CODE REVIEW

**Comment by alangenfeld** [comment]
Putting this up early to give @bengotow a preview.

**Comment by alangenfeld** [comment]
Screen Shot 2019-04-11 at 11 32 22 AM

File: python_modules/dagster/dagster/core/definitions/repository.py [hunk 1]
```diff
@@ -162,3 +175,96 @@ def solid_def_named(self, name):
...
+
+    def get_preset(self, pipeline_name, solid_name):
+        check.str_param(pipeline_name, 'pipeline_name')
+        check.str_param(solid_name, 'solid_name')
+        return self.get_presets_for_pipeline(pipeline_name).get(solid_name)
+
+    def get_presets_for_pipeline(self, pipeline_name):
+        if self.repo_config and self.repo_config.get('pipelines'):
+            return {
+                name: PipelinePreset(
```
Comment (schrockn) [inline python_modules/dagster/dagster/core/definitions/repository.py]: this seems a bit too clever by half. It's not obvious what the shape of this is at first read. Dict[str, List[PipelinePresent]] ?

File: python_modules/dagster/dagster/check/__init__.py [hunk 2]
```diff
@@ -309,6 +309,22 @@ def opt_list_param(obj_list, param_name, of_type=None):
     return _check_list_items(obj_list, of_type)
 
 
+def opt_nullable_list_param(obj_list, param_name, of_type=None):
```
Comment (schrockn) [inline python_modules/dagster/dagster/check/__init__.py]: add some tests to these to demonstrate different behavior,  get code coverage, etc

File: python_modules/dagster/dagster/check/__init__.py [hunk 3]
```diff
@@ -371,6 +387,20 @@ def opt_dict_param(obj, param_name, key_type=None, value_type=None, value_class=
     return _check_key_value_types(obj, key_type, value_type)
 
 
+def opt_nullable_dict_param(obj, param_name, key_type=None, value_type=None, value_class=None):
```
Comment (schrockn) [inline python_modules/dagster/dagster/check/__init__.py]: samesies

**Comment by schrockn** [comment]
Resolves https://github.com/dagster-io/dagster/issues/244

**Pull request merged**